### PR TITLE
8293287 add ReplayReduce flag

### PR DESCRIPTION
**Pull request by dean-long**
Add an experimental flag to help developers "reduce" a replay file.

As a first step, I plan to simulate reduced inlining. This will output multiple "compile" lines as if the first level of inlining never happened:
A --> B --> C
A --> D --> E
becomes
B --> C
D --> E
Developers can repeat iteratively until the replay crash no longer reproduces.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8293287](https://bugs.openjdk.org/browse/JDK-8293287): add ReplayReduce flag

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to [73d6e812](https://git.openjdk.org/jdk/pull/10134/files/73d6e8128cd7f4881659f3d76541dab10badc767)
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10134/head:pull/10134` \
`$ git checkout pull/10134`

Update a local copy of the PR: \
`$ git checkout pull/10134` \
`$ git pull https://git.openjdk.org/jdk pull/10134/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10134`

View PR using the GUI difftool: \
`$ git pr show -t 10134`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10134.diff">https://git.openjdk.org/jdk/pull/10134.diff</a>

</details>

### CODE REVIEW

**Comment by dean-long** [comment]
One test is failing -- moving back to DRAFT.

File: src/hotspot/share/opto/compile.cpp [hunk 1]
```diff
@@ -4573,6 +4573,33 @@ void Compile::dump_inline_data(outputStream* out) {
   }
 }
 
+void Compile::dump_inline_data_reduced(outputStream* out) {
```
Comment (vnkozlov) [inline src/hotspot/share/opto/compile.cpp:4576]: Add `assert(ReplayReduce`.

File: src/hotspot/share/ci/ciEnv.cpp [hunk 2]
```diff
@@ -1647,6 +1647,12 @@ void ciEnv::find_dynamic_call_sites() {
 void ciEnv::dump_compile_data(outputStream* out) {
   CompileTask* task = this->task();
   if (task) {
+#ifdef COMPILER2
+    if (ReplayReduce && compiler_data() != NULL) {
```
Comment (vnkozlov) [inline src/hotspot/share/ci/ciEnv.cpp:1651]: Is this feature when you are replaying compilation? If yes, add `ReplayCompiles` check too.

Comment (dean-long) [reply src/hotspot/share/ci/ciEnv.cpp:1651]: It seems most useful with ReplayCompiles, but I don't want to require it.  A developer might want to turn on ReplayReduce without ReplayCompiles, to save a replay step, for example when running generated tests from creduce or javafuzzer without an existing replay file.

**Review by vnkozlov — APPROVED** [review]
Okay.

**Comment by dean-long** [comment]
Unfortunately, the TestVMNoCompLevel.java test is still failing on some platforms.  I'm tempted to remove that test, as it doesn't not seem to add value anymore.

**Comment by vnkozlov** [comment]
> Unfortunately, the TestVMNoCompLevel.java test is still failing on some platforms. I'm tempted to remove that test, as it doesn't not seem to add value anymore.

Can you explain the issue with this test and your changes? New code should be off by default.

**Comment by dean-long** [comment]
> Can you explain the issue with this test and your changes? New code should be off by default.

I had to move the call to reset() (which clears parsing errors) to allow multiple compile commands.  This revealed an existing problem with compiler/ciReplay/TestVMNoCompLevel.java.  This test removes the last token from the "compile" line, trying to simulate a pre-2013 replay file.  Removing the last token is wrong if there are inlining tokens.  I could fix the test, but I don't see the value.  I don't think a 2022 JVM needs to read pre-2013 replay files.  If necessary, an old replay file can be edited to allow it to be parsed by a recent JVM.  I proposed to deleted the test and related workarounds in the code.

**Review by vnkozlov — APPROVED** [review]
Okay

File: src/hotspot/share/opto/compile.cpp [hunk 3]
```diff
@@ -4573,6 +4573,35 @@ void Compile::dump_inline_data(outputStream* out) {
...
+  assert(ReplayReduce, "");
+
+  InlineTree* inl_tree = ilt();
+  if (inl_tree == NULL) {
+    return;
+  }
+  // Enable interative replay file reduction
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/compile.cpp:4583]:
```suggestion
  // Enable iterative replay file reduction
```

**Review by TobiHartmann — COMMENTED** [review]
I'm wondering if this functionality should really be part of the VM. Wouldn't a simple script that regex-parses the compile statement of the replay file, iteratively removes inlines and runs replay compilation to check if the issue still reproduces, be more powerful and easier to maintain? It could be combined with also removing class loading statements.

**Comment by dean-long** [comment]
> I'm wondering if this functionality should really be part of the VM. Wouldn't a simple script that regex-parses the compile statement of the replay file, iteratively removes inlines and runs replay compilation to check if the issue still reproduces, be more powerful and easier to maintain? It could be combined with also removing class loading statements.

Writing a script was my first attempt, but getting the parsing right would have taken more time than I wanted to invest, so I did what was quickest and easiest.  I found it useful, so I thought checking it in might be useful to others, considering the code changes are small.

One problem with an external script is that replay won't normally create a new replay file if  -XX:+ReplayCompiles is on.  So, some kind of JVM change is required.

**Review by TobiHartmann — APPROVED** [review]
Okay, I'm fine with adding this functionality, if it's been helpful to you.

**Comment by dean-long** [comment]
Thanks Tobias and Vladimir.

**Pull request closed without merging**